File: android/app/src/main/java/com/focus/v2android/core/GazeTracker.kt
package com.focus.v2android.core

import android.os.SystemClock
import kotlin.math.hypot
import kotlin.math.sqrt

/**
 * Python core/gaze_tracker.py 포트.
 * I-VT saccade + I-DT fixation + EAR 계산.
 * 입력: MediaPipe FaceLandmarker 정규화 랜드마크 (478*2 flattened FloatArray)
 */
class GazeTracker(
    private val calibration: CalibrationModel,
    private val saccadeVelocityPxS: Float = 1500f,
    private val fixationDispersionPx: Float = 100f,
    private val fixationMinDurationS: Float = 0.12f,
    private val dispersionWindow: Int = 10,
    private val smoothingAlpha: Float = 0.35f
) {
    // EMA 상태
    private var smoothedX: Float? = null
    private var smoothedY: Float? = null
    private var prevSample: GazeSample? = null

    // Fixation 상태
    private var fixationStart: Long? = null
    private val fixationWindow = ArrayDeque<Triple<Long, Float, Float>>(dispersionWindow)

    fun reset() {
        smoothedX = null; smoothedY = null
        prevSample = null
        fixationStart = null
        fixationWindow.clear()
    }

    /**
     * 478개 정규화 랜드마크(flattened xy)에서 GazeSample 생성.
     * Python process_with_landmarks()와 동일.
     */
    fun processWithLandmarks(landmarks: FloatArray): GazeSample? {
        val (fx, fy) = calibration.extractGazeFeature(landmarks) ?: run {
            reset()
            return null
        }

        val now = SystemClock.elapsedRealtime()

        // 화면 좌표 EMA 스무딩
        val rawX: Float
        val rawY: Float
        val (rx, ry) = calibration.apply(fx, fy)
        rawX = rx; rawY = ry

        val alpha = smoothingAlpha.coerceIn(0.05f, 1f)
        smoothedX = if (smoothedX == null) rawX else alpha * rawX + (1f - alpha) * smoothedX!!
        smoothedY = if (smoothedY == null) rawY else alpha * rawY + (1f - alpha) * smoothedY!!
        val sx = smoothedX!!; val sy = smoothedY!!

        // 속도 계산
        val velocity = prevSample?.let { prev ->
            val dt = maxOf((now - prev.timestampMs) / 1000f, 0.001f)
            hypot(sx - prev.screenX, sy - prev.screenY) / dt
        } ?: 0f

        val isSaccade = velocity > saccadeVelocityPxS
        val (inFix, fixDur) = updateFixation(now, sx, sy, isSaccade)

        // EAR
        val ear = calibration.computeMeanEar(landmarks)

        val sample = GazeSample(
            timestampMs = now,
            featureX = fx, featureY = fy,
            screenX = sx, screenY = sy,
            velocityPxPerS = velocity,
            isSaccade = isSaccade,
            inFixation = inFix,
            fixationDurationS = fixDur,
            ear = ear
        )
        prevSample = sample
        return sample
    }

    private fun updateFixation(nowMs: Long, sx: Float, sy: Float, isSaccade: Boolean): Pair<Boolean, Float> {
        if (isSaccade) {
            fixationStart = null
            fixationWindow.clear()
            return Pair(false, 0f)
        }
        if (fixationWindow.size >= dispersionWindow) fixationWindow.removeFirst()
        fixationWindow.addLast(Triple(nowMs, sx, sy))

        val xs = fixationWindow.map { it.second }
        val ys = fixationWindow.map { it.third }
        val dispersion = (xs.max()!! - xs.min()!!) + (ys.max()!! - ys.min()!!)

        if (dispersion <= fixationDispersionPx) {
            if (fixationStart == null) fixationStart = fixationWindow.first().first
            val durS = (nowMs - fixationStart!!) / 1000f
            return Pair(durS >= fixationMinDurationS, durS)
        }
        fixationStart = null
        return Pair(false, 0f)
    }
}

File: src/lib/focus/gazeTracker.ts
/**
 * focus/gazeTracker.ts — core/GazeTracker.kt 의 TS 포트.
 * I-VT saccade + I-DT fixation + EAR. 시간은 performance.now() ms 사용.
 */
import type { CalibrationModel } from './calibrationModel';
import type { GazeSample } from './types';

export class GazeTracker {
    private smoothedX: number | null = null;
    private smoothedY: number | null = null;
    private prevSample: GazeSample | null = null;
    private fixationStart: number | null = null;
    private fixationWindow: Array<[number, number, number]> = []; // [t, x, y]

    private calibration: CalibrationModel;
    private saccadeVelocityPxS: number;
    private fixationDispersionPx: number;
    private fixationMinDurationS: number;
    private dispersionWindow: number;
    private smoothingAlpha: number;

    constructor(
        calibration: CalibrationModel,
        saccadeVelocityPxS = 1500,
        fixationDispersionPx = 100,
        fixationMinDurationS = 0.12,
        dispersionWindow = 10,
        smoothingAlpha = 0.35,
    ) {
        this.calibration = calibration;
        this.saccadeVelocityPxS = saccadeVelocityPxS;
        this.fixationDispersionPx = fixationDispersionPx;
        this.fixationMinDurationS = fixationMinDurationS;
        this.dispersionWindow = dispersionWindow;
        this.smoothingAlpha = smoothingAlpha;
    }

    reset(): void {
        this.smoothedX = null; this.smoothedY = null;
        this.prevSample = null;
        this.fixationStart = null;
        this.fixationWindow = [];
    }

    processWithLandmarks(landmarks: Float32Array, nowMs: number): GazeSample | null {
        const feat = this.calibration.extractGazeFeature(landmarks);
        if (!feat) { this.reset(); return null; }
        const [fx, fy] = feat;

        const [rawX, rawY] = this.calibration.apply(fx, fy);
        const alpha = Math.min(Math.max(this.smoothingAlpha, 0.05), 1);
        this.smoothedX = this.smoothedX == null ? rawX : alpha * rawX + (1 - alpha) * this.smoothedX;
        this.smoothedY = this.smoothedY == null ? rawY : alpha * rawY + (1 - alpha) * this.smoothedY;
        const sx = this.smoothedX, sy = this.smoothedY;

        let velocity = 0;
        if (this.prevSample) {
            const dt = Math.max((nowMs - this.prevSample.timestampMs) / 1000, 0.001);
            velocity = Math.hypot(sx - this.prevSample.screenX, sy - this.prevSample.screenY) / dt;
        }
        const isSaccade = velocity > this.saccadeVelocityPxS;
        const [inFix, fixDur] = this.updateFixation(nowMs, sx, sy, isSaccade);
        const ear = this.calibration.computeMeanEar(landmarks);

        const sample: GazeSample = {
            timestampMs: nowMs,
            featureX: fx, featureY: fy,
            screenX: sx, screenY: sy,
            velocityPxPerS: velocity,
            isSaccade, inFixation: inFix,
            fixationDurationS: fixDur,
            ear,
        };
        this.prevSample = sample;
        return sample;
    }

    private updateFixation(nowMs: number, sx: number, sy: number, isSaccade: boolean): [boolean, number] {
        if (isSaccade) {
            this.fixationStart = null;
            this.fixationWindow = [];
            return [false, 0];
        }
        if (this.fixationWindow.length >= this.dispersionWindow) this.fixationWindow.shift();
        this.fixationWindow.push([nowMs, sx, sy]);

        let minX = Infinity, maxX = -Infinity, minY = Infinity, maxY = -Infinity;
        for (const [, wx, wy] of this.fixationWindow) {
            if (wx < minX) minX = wx; if (wx > maxX) maxX = wx;
            if (wy < minY) minY = wy; if (wy > maxY) maxY = wy;
        }
        const dispersion = (maxX - minX) + (maxY - minY);
        if (dispersion <= this.fixationDispersionPx) {
            if (this.fixationStart == null) this.fixationStart = this.fixationWindow[0][0];
            const durS = (nowMs - this.fixationStart) / 1000;
            return [durS >= this.fixationMinDurationS, durS];
        }
        this.fixationStart = null;
        return [false, 0];
    }
}
